File: packages/pythonlib/src/main/python/service_manager/service_manager.py
"""
This module defines types and functions related to a service manager, 
which is used by clients to retrieve service instances
"""

from service_configuration import EMPTY_SERVICE_CONFIGURATION
from service_manager_strategy import (
    ServiceManagerStrategy,
    ConfigServiceManagerStrategy,
)
from service import TService, CConfiguration


class ServiceManager:
    """
    The ServiceManager is used to manage and retrieve services.  It's a proxy to
    a ServiceManagerStrategy implementation, which is used to modify the
    ServiceManager behavior.  The static proxy simplifies client code
    """

    _service_manager_strategy: ServiceManagerStrategy = ConfigServiceManagerStrategy()

    @classmethod
    def getService[
        TService, CConfiguration
    ](name: str, config: CConfiguration = EMPTY_SERVICE_CONFIGURATION) -> TService:
        """
        Retrieve a service.

        An optional config may be supplied for service instance level
        configuration.  This must be supported by the associated
        ServiceLifecycleManager
        """
        return ServiceManager._service_manager_strategy.get_service[TService](
            name, config
        )

    @classmethod
    def is_service_defined(cls, name: str) -> bool:
        """
        Determine if a service with the specified name has been defined
        """
        return ServiceManager._service_manager_strategy.is_service_defined(name)

    @classmethod
    def shutdown(cls) -> None:
        """
        Shutdown the ServiceManager and all associated services
        """
        ServiceManager._service_manager_strategy.shutdown()

    @classmethod
    def setDefaultStrategy(cls, strategyToSet: ServiceManagerStrategy) -> None:
        """
        Set the service manager strategy.  This strategy will be used to retrieve services based on the algorithm defined
        in the ServiceManagerFactory implementnatation

        In the future, a ServiceManager may be able to use multiple strategies concurrently.  As of the current API, it
        only supports a single strategy

        @param strategyToSet the service manager strategy to set
        """
        ServiceManager._service_manager_strategy = strategyToSet

File: packages/javalib/src/main/java/com/sphyrna/servicemanager/ServiceManager.java
package com.sphyrna.servicemanager;

import com.sphyrna.servicemanager.providers.ConfigServiceManagerStrategy;
import com.sphyrna.servicemanager.providers.MapConfiguration;
import java.util.Objects;

/**
 * The ServiceManager is used to manage and retrieve services
 *
 * @author sgoldstein
 *
 */
public class ServiceManager
{
    private static ServiceManagerStrategy serviceManagerStrategy = new ConfigServiceManagerStrategy();

    /**
     * Do not instantiate
     */
    private ServiceManager() {}

    /**
     * Retrieve a service
     *
     * @param name
     *            the name of the service
     * @return the service associated with the specified name
     * @throws ServiceNotAvailableException
     *             if the specified service does not exist
     */
    public static <S extends Service<ServiceConfiguration>> S getService(String name)
        throws ServiceNotAvailableException, ServiceException
    {
        Objects.requireNonNull(name, "name cannot be null");

        return serviceManagerStrategy.getService(name);
    }

    /**
     * Retrieve a service
     *
     * @param name
     *            the name of the service
     * @param serviceConfig
     * @return the service associated with the specified name
     * @throws ServiceNotAvailableException
     *             if the specified service does not exist
     */
    public static <S extends Service<C>, C extends ServiceConfiguration> S getService(String name, C serviceConfig)
        throws ServiceNotAvailableException, ServiceException
    {
        Objects.requireNonNull(name, "name cannot be null");
        Objects.requireNonNull(serviceConfig, "serviceConfig cannot be null");

        return serviceManagerStrategy.getService(name, serviceConfig);
    }

    /**
     * Determine if a service with the specified name is defined
     *
     * @param name
     *            the name of the service to investigate
     * @return true if the service is available; false otherwise
     */
    public static boolean isServiceDefined(String name)
    {
        Objects.requireNonNull(name, "name cannot be null");

        return serviceManagerStrategy.isServiceDefined(name);
    }

    /**
     * Shutdown the ServiceManager and all associated services
     */
    public static void shutdown()
    {
        serviceManagerStrategy.shutdown();
    }

    /**
     * Set the service manager strategy.  This strategy will be used to retrieve services based on the algorithm defined
     * in the ServiceManagerFactory implementnatation
     *
     * In the future, a ServiceManager may be able to use multiple strategies concurrently.  As of the current API, it
     * only supports a single strategy
     *
     * @param strategyToSet the service manager strategy to set
     */
    public static void setDefaultStrategy(ServiceManagerStrategy strategyToSet)
    {
        serviceManagerStrategy = Objects.requireNonNull(strategyToSet, "strategyToSet cannot be null");
    }
}
